modules/cluster_estimation/cluster_estimation.py

```diff
@@ -0,0 +1,342 @@
+        """
+        diff_vector = [point[0] - cluster[0], point[1] - cluster[1]]
+
+        return np.linalg.norm(diff_vector)
+
+    @staticmethod
+    def __convert_detections_to_point(detections: "list[DetectionInWorld]") \
+        -> "list[tuple[float, float]]":
+        """
+        Convert DetectionInWorld input object to a list of points- (x,y) positions, to store.
+
+        PARAMETERS
+        ----------
+        detections: list[DetectionInWorld]
+            List of DetectionInWorld intermediate objects, the data structure that is passed to the
+            worker
+
+        RETURNS
+        -------
+        points: list[tuple[float, float]]
+            List of points (x,y)
+        -------
+        """
+        points = []
+        for detection in detections:
+            # attribute centre holds positioning data
+            points.append(tuple([detection.centre[0], detection.centre[1]]))
+
+        return points
+
+    def __filter_by_points_ownership(self,
+                                     model_output: "list[np.ndarray, float, float]") \
+        -> "list[np.ndarray, float, float]":
+        """
+        Removes any clusters that don't have any points belonging to it.
+
+        PARAMETERS
+        ----------
+        model_output: list[np.ndarray, float, float]
+            List containing predicted cluster centers, with each element having the format
+            [(x position, y position), weight, covariance)].
+
+        RETURNS
+        -------
+        filtered_output: list[np.ndarray, float, float]
+            List containing predicted cluster centers after filtering.
+        """
+
+        results = self.__vgmm.predict(self.__all_points)
+        filtered_output = []
+
+        # Filtering by each cluster's point ownership
+        unique_clusters, num_points_per_cluster = np.unique(results, return_counts=True)
+        # Remove empty clusters
+        i = 0
+        for i in range(len(model_output)):
+            if i in unique_clusters:
+                filtered_output.append(model_output[i])
+
+        return filtered_output
+
+    def __filter_by_covariances(self, model_output: list[np.ndarray, float, float]) \
+        -> "list[np.ndarray, float, float]":
+        """
+        Removes any cluster with covariances much higher than the lowest covariance value.
+
+        PARAMETERS
+        ----------
+        model_output: list[np.ndarray, float, float]
+            List containing predicted cluster centers, with each element having the format
+            [(x position, y position), weight, covariance)].
+
+        RETURNS
+        -------
+        list[np.ndarray, float, float]
+            List containing predicted cluster centers after filtering by covariance.
+        """
+        # python list and not np array, need to loop through manually
+        min_covariance = float("inf") 
+        for item in model_output:
+            if item[2] < min_covariance:
+                min_covariance = item[2]
```

Comment:
I'm not gonna lie, I have no idea how to use min() in this case. The min(model_output)[2] didn't work, and since the data structure is list[tuple[ _ , weights, _ ]] I can't specify I want weights for the min() to work.

Reply:
Seems like where I got it from is spreading misinformation, unfortunate.